src/uct/ugni/base/ugni_ep.c

```diff
@@ -6,43 +6,111 @@
+                                                  ucs_arbiter_elem_t *elem,
+                                                  void *arg){
+  uct_pending_req_t *req = ucs_container_of(elem, uct_pending_req_t, priv);
+  ucs_status_t rc;
+  
+  rc = req->func(req);
+  ucs_trace_data("progress pending request %p returned %s", req,
+                 ucs_status_string(rc));
+
+
+  if (rc == UCS_OK) {
+    /* sent successfully. remove from the arbiter */
+    return UCS_ARBITER_CB_RESULT_REMOVE_ELEM;
+  } else {
+    /* couldn't send. keep this request in the arbiter until the next time
+     * this function is called */
+    return UCS_ARBITER_CB_RESULT_RESCHED_GROUP;
+  }
+}
+
+static ucs_arbiter_cb_result_t uct_ugni_ep_abriter_purge_cb(ucs_arbiter_t *arbiter,
+                                                           ucs_arbiter_elem_t *elem,
+                                                           void *arg){
+  uct_pending_req_t *req = ucs_container_of(elem, uct_pending_req_t, priv);
+  uct_pending_callback_t cb = arg;
+
+  cb(req);
+  return UCS_ARBITER_CB_RESULT_REMOVE_ELEM;
+}
+
+void uct_ugni_ep_pending_purge(uct_ep_h tl_ep, uct_pending_callback_t cb){
+  uct_ugni_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_ugni_iface_t);
+  uct_ugni_ep_t *ep = ucs_derived_of(tl_ep, uct_ugni_ep_t);
+
+  ucs_arbiter_group_purge(&iface->arbiter, &ep->arb_group,
+                          uct_ugni_ep_abriter_purge_cb, cb);
+}
+
+ucs_status_t ugni_connect_ep(uct_ugni_iface_t *iface, const uct_sockaddr_ugni_t *iface_addr, uct_ugni_ep_t *ep){
+  gni_return_t ugni_rc;
+
+  ugni_rc = GNI_EpBind(ep->ep, iface_addr->nic_addr, iface_addr->domain_id);
+  if (GNI_RC_SUCCESS != ugni_rc) {
+    (void)GNI_EpDestroy(ep->ep);
+    ucs_error("GNI_EpBind failed, Error status: %s %d",
+              gni_err_str[ugni_rc], ugni_rc);
+    return UCS_ERR_UNREACHABLE;
+  }
+
+  ucs_debug("Binding ep %p to address (%d %d)", ep, iface_addr->nic_addr,
+            iface_addr->domain_id);
+
+  ep->outstanding = 0;
+
+  return UCS_OK;
+}
+
 /* Endpoint definition */
 UCS_CLASS_INIT_FUNC(uct_ugni_ep_t, uct_iface_t *tl_iface,
                     const struct sockaddr *addr)
 {
     uct_ugni_iface_t *iface = ucs_derived_of(tl_iface, uct_ugni_iface_t);
     const uct_sockaddr_ugni_t *iface_addr = (const uct_sockaddr_ugni_t*)addr;
+    ucs_status_t rc = UCS_OK;
     gni_return_t ugni_rc;
 
-    UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super)
+    UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super);
 
     ugni_rc = GNI_EpCreate(iface->nic_handle, iface->local_cq, &self->ep);
     if (GNI_RC_SUCCESS != ugni_rc) {
-        ucs_error("GNI_CdmCreate failed, Error status: %s %d",
-                  gni_err_str[ugni_rc], ugni_rc);
-        return UCS_ERR_NO_DEVICE;
+      ucs_error("GNI_CdmCreate failed, Error status: %s %d",
+                gni_err_str[ugni_rc], ugni_rc);
+      return UCS_ERR_NO_DEVICE;
```

Comment:
M-C-\ is the macro.

Which reminds me a I need to make a PR for .gitignore to ignore some emacs files.